trieste/models/keras/interface.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2021 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import ABC, abstractproperty
+from typing import Optional, Union
+
+import tensorflow as tf
+
+from ...types import TensorType
+from ..interfaces import ProbabilisticModel
+from ..optimizer import BatchOptimizer
+
+
+class NeuralNetworkPredictor(ProbabilisticModel, tf.Module, ABC):
```

Comment:
Maybe this should be `KerasPredictor`, that seems to be a higher level name, i.e. we might have keras models that are not NNs?

Reply:
If we request an optimizer, then it should be a `TrainableProbabilisticModel`, right?

Reply:
@henrymoss yes, we can have neural network models in TF that are not Keras, Keras is just a higher level framework that facilitates building them, but interface should be flexible enough that someone could use a model that is not Keras - many researchers would use this, as for developing new types of neural nets most likely you cannot do it with Keras

Reply:
But, suppose we wanted to write a wrapper for a Keras model that wasnt a NN. Then it would be silly to use this class name?

Reply:
@vpicheny I think it makes sense that optimizer is defined here because all underlying models will need to have it so we might as well standardize it with an interface

Reply:
> But, suppose we wanted to write a wrapper for a Keras model that wasnt a NN. Then it would be silly to use this class name?

Keras focuses on deep learning, I don't think it makes sense to use it for other types of models, in any case the idea here is to support neural network types of models, using Keras is just for convenience

perhaps this goes back to what we should call the supported model classes - GPs, DGPs and NNs rather than gpflow, gpflux and keras

Reply:
you're right, better here to copy what we have for gpflow and solve this later.

Reply:
On @henrymoss comment: I think the question was: could we have keras models that are not NNs? in that case, wouldn't KerasPredictor be a better name? So this interface ties with keras and not NNs. I don't see anything NN specific here, except perhaps the restriction on predict joint.

Reply:
@vpicheny there was another comment above that you might have missed?


> > But, suppose we wanted to write a wrapper for a Keras model that wasnt a NN. Then it would be silly to use this class name?
> 
> Keras focuses on deep learning, I don't think it makes sense to use it for other types of models, in any case the idea here is to support neural network types of models, using Keras is just for convenience
> 
> perhaps this goes back to what we should call the supported model classes - GPs, DGPs and NNs rather than gpflow, gpflux and keras